packages/nextcloud_test/lib/src/test_target/test_target.dart

```diff
@@ -0,0 +1,75 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'package:cookie_store/cookie_store.dart';
+import 'package:neon_http_client/neon_http_client.dart';
+import 'package:nextcloud/nextcloud.dart';
+import 'package:nextcloud_test/nextcloud_test.dart';
+import 'package:nextcloud_test/src/fixtures.dart';
+import 'package:nextcloud_test/src/proxy_http_client.dart';
+import 'package:nextcloud_test/src/test_target/docker_container.dart';
+import 'package:nextcloud_test/src/test_target/local.dart';
+import 'package:version/version.dart';
+
+/// Factory for creating [TestTargetInstance]s.
+abstract interface class TestTargetFactory<T extends TestTargetInstance> {
+  /// Creates a new [TestTargetFactory].
+  static TestTargetFactory create() {
+    final url = Platform.environment['URL'];
+    final dir = Platform.environment['DIR'];
+    if (url != null && url.isNotEmpty && dir != null && dir.isNotEmpty) {
+      return LocalFactory();
+    }
+
+    return DockerContainerFactory();
+  }
+
+  /// Spawns a new [T].
+  FutureOr<T> spawn(Preset? preset);
+
+  /// Returns the available presets for the factory.
+  Map<String, List<Version>> getPresets();
+}
+
+/// Instance of a test target.
+abstract class TestTargetInstance {
+  /// Destroys the instance.
+  FutureOr<void> destroy();
+
+  /// URL where the target is available from the host side.
+  Uri get hostURL;
+
+  /// URL where the target is available from itself.
+  Uri get targetURL;
+
+  /// Creates an app password for [username] on the instance.
+  FutureOr<String> createAppPassword(String username);
+
+  /// Creates a new [NextcloudClient] for a given [username].
+  ///
+  /// It is expected that the password of the user matches the its [username].
+  FutureOr<NextcloudClient> createClient({
```

Comment:
This always returns a future
```suggestion
  Future<NextcloudClient> createClient({
```